example-client/client.py

```diff
@@ -64,43 +64,76 @@ def run_structure(input: str) -> Optional[str]:
         args=[input],
     )
 
-    # Runs are asynchronous, so we need to poll the status until it's no longer running.
     structure_run_id = structure_run["structure_run_id"]
-    status = structure_run["status"]
-    printed_event_ids = set()  # Keep track of which events we've printed.
-    while status not in ("SUCCEEDED", "FAILED"):
-        structure_run = get_structure_run(
-            host=HOST, api_key=GT_API_KEY, run_id=structure_run_id
-        )
-        status = structure_run["status"]
 
-        # You can comment out this block if you don't want to stream events as they occur.
+    received_event_ids = set()  # Keep track of which events we've already received.
+    structure_has_terminal_event = False
+
+    # The client will continuously poll the Structure Run until it receives one of three events:
+    # 1. Our Structure has voluntarily emitted a FinishStructureRunEvent
+    # 2. The Structure has stopped running due to an error
+    # 3. The Structure has stopped running without an error
+    while not structure_has_terminal_event:
+        # TODO: Can the API be adjusted to give us all NEW events that occurred after a specific point?
         event_list = get_structure_run_events(
             host=HOST, api_key=GT_API_KEY, run_id=structure_run_id
         )
         events = event_list["events"]
-        printed_event_ids = print_streaming_events(events, printed_event_ids)
 
-        time.sleep(1)  # Poll every second.
+        # Find all events that are new to us.
+        for event in events:
+            event_id = event["event_id"]
+            if event_id not in received_event_ids:
+                # Mark it as received.
+                received_event_ids.add(event_id)
+
+                # Is this an event type that we care about?
+                event_type = event["value"]["type"]
```

Comment:
We should not use the framework event type here. This should be keying off a top level event type field set by Cloud.

Reply:
can you give an example of what you're thinking? I want to make sure we're not foisting more work on the client, and maybe we're not.

Reply:
`event["value"]["type"]` is a framework-specific field. Each of the "Cloud Events" should contain its own top level `type` field.  

```python
event_type = event["type"]

match event_type:
    case "UserEvent.CompletionChunkEvent":
        # If the agent is set to streaming, print out each streaming chunk
        completion_token = event["value"]["token"]
        print(completion_token, flush=True, end="")
    case "UserEvent.FinishStructureRunEvent":
    ...
    case "SystemEvent.StructureRunExecutionFailed":
    ...
```

Reply:
oh got it, embedding it in the name of the event type. I was concerned we were going to add a second `type` field at the top level of `event`, which would have forced more boilerplate for the client. Totally onboard with your suggestion.

Reply:
@SavagePencil I did add a `type` field at the top level of `event`. But yes it includes the framework type event too. Just to entertain what would happen if we didn't...
```python
event_type = event["type"]

match event_type:
    case "UserEvent":
       framework_event_type = event["value"]["type"]
       match framework_event_type:
           case "CompletionChunkEvent":      
               # If the agent is set to streaming, print out each streaming chunk
              completion_token = event["value"]["token"]
              print(completion_token, flush=True, end="")
           case "FinishStructureRunEvent":
           ...
    case "SystemEvent.StructureRunExecutionFailed":
    ...
```

Putting it in the name really only works because we know the event is coming from the framework and we know to look at `event["value"]["type"]`. If we wanted to allow for non-framework user events we'd either need to enforce a similar structure or do something like this ^.

Reply:
bleh, I think customers would much prefer the dotted notation inside the string than having to juggle those. I feel like we're doing this for OUR convenience over theirs. Happy to get one of our Jasons to settle this for us.

Reply:
I'm cool with namespacing in the event name. Main takeaway is that there is a Cloud-inserted top level event type:
```python
event_type = event["type"]
```

Reply:
@collindutter how do you feel about being more explicit to make it clear when an event came from the service ("system" feels ambiguous to me). e.g., `GriptapeCloudEvent`?

Reply:
But then I'd ask what _kind_ of Griptape Cloud Event is it? Is it one the "system" generated, or one I, the "user", generated?

Reply:
Hrmph, it worked for *me* 😤. I was hoping to capture the essence of "this event happened outside of my code, out of my control." Any terms you can think of? Should we crack open the thesaurus?